File: frontend/src/api/sources.ts
import type { components } from './generated/openapi'
import { ApiError, requestJson } from './client'
import { hasText, isRecord, isUuid } from './json-guards'

export type SourceDto = components['schemas']['SourceResponse']

export async function listSources(signal?: AbortSignal): Promise<SourceDto[]> {
  const response = await requestJson<unknown>('/sources', { method: 'GET', signal })
  if (!Array.isArray(response) || !response.every(isSource)) {
    throw invalidResponse()
  }
  return response
}

export async function bindSource(id: string, knowledgeBaseId: string | null): Promise<SourceDto> {
  const response = await requestJson<unknown>(`/sources/${encodeURIComponent(id)}/knowledge-base`, {
    method: 'PATCH',
    body: JSON.stringify({ knowledge_base_id: knowledgeBaseId }),
  })
  if (!isSource(response)) throw invalidResponse()
  return response
}

function isSource(value: unknown): value is SourceDto {
  return (
    isRecord(value) &&
    isUuid(value.id) &&
    hasText(value.provider) &&
    hasText(value.external_id) &&
    hasText(value.name) &&
    (value.feed_url === null || typeof value.feed_url === 'string') &&
    (value.home_url === null || typeof value.home_url === 'string') &&
    (value.knowledge_base_id === null || isUuid(value.knowledge_base_id)) &&
    (value.knowledge_base_key === null || hasText(value.knowledge_base_key)) &&
    (value.sync_checkpoint === null || hasText(value.sync_checkpoint)) &&
    (value.sync_checkpoint_updated_at === null ||
      (typeof value.sync_checkpoint_updated_at === 'string' &&
        !Number.isNaN(Date.parse(value.sync_checkpoint_updated_at))))
  )
}

function invalidResponse(): ApiError {
  return new ApiError({
    message: 'The source service returned an invalid response.',
    code: 'response_invalid',
  })
}

File: backend/src/agent_lab/schemas/sources.py
"""Source 配置列表和 KnowledgeBase 绑定请求的 HTTP 契约。"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class SourceKnowledgeBaseBindingRequest(BaseModel):
    """把来源绑定到一个目标库；null 表示解除绑定。"""

    knowledge_base_id: UUID | None = Field(description="目标 KnowledgeBase ID；null 表示解除绑定。")

    model_config = ConfigDict(extra="forbid", frozen=True, hide_input_in_errors=True)


class SourceResponse(BaseModel):
    """来源元数据、同步状态和当前绑定状态。"""

    id: UUID
    provider: str
    external_id: str
    name: str
    feed_url: str | None
    home_url: str | None
    knowledge_base_id: UUID | None
    knowledge_base_key: str | None
    sync_checkpoint: str | None
    sync_checkpoint_updated_at: datetime | None

    model_config = ConfigDict(from_attributes=True, frozen=True)


class SourceErrorResponse(BaseModel):
    """来源绑定失败的稳定脱敏错误。"""

    code: str
    detail: str
    retryable: bool

    model_config = ConfigDict(frozen=True)
